dwave_networkx/algorithms/TSP.py

```diff
@@ -0,0 +1,157 @@
+
+    A route is a cycle in the graph that reaches each node exactly once. 
+    A minimum route is a route with the smallest total edge weight.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+        The graph on which to find a maximum cut weighted independent set.
+
+    sampler
+        A binary quadratic model sampler. A sampler is a process that
+        samples from low energy states in models defined by an Ising
+        equation or a Quadratic Unconstrained Binary Optimization
+        Problem (QUBO). A sampler is expected to have a 'sample_qubo'
+        and 'sample_ising' method. A sampler is expected to return an
+        iterable of samples, in order of increasing energy. If no
+        sampler is provided, one must be provided using the
+        `set_default_sampler` function.
+        
+    lagrange : optional (default 2)
+        Lagrange parameter to weight constraints (no edges within set) 
+        versus objective (largest set possible).
+
+    sampler_args
+        Additional keyword parameters are passed to the sampler.
+
+    Returns
+    -------
+    route : list
+       List of nodes in order to be visited on a route
+
+    Notes
+    -----
+    Samplers by their nature may not return the optimal solution. This
+    function does not attempt to confirm the quality of the returned
+    sample.
+
+    References
+    ----------
+
+    """
+    # Get a QUBO representation of the problem
+    Q = traveling_salesman_qubo(G, lagrange)
+
+    # use the sampler to find low energy states
+    response = sampler.sample_qubo(Q, **sampler_args)
+
+    # we want the lowest energy sample
+    sample = next(iter(response))
+
+    # Define route for solution found
+    N = len(G)
+    route = [-1]*N
+    route[0]=0
+    for node in sample:
+        if sample[node]>0:
+            j = node%N
+            v = (node-j)/N
+            route[j] = int(v)
+
+    # nodes that are spin up or true are exactly the ones in S.
+    return route
+
+def traveling_salesman_qubo(G, lagrange=2.0):
+    """Return the QUBO with ground states corresponding to a minimum TSP route.
+    Parameters
+    ----------
+    G : NetworkX graph
+
+        Nodes in graph must be labeled 0...N-1
+        
+    lagrange : optional (default 2)
+        Lagrange parameter to weight constraints (no edges within set) 
+        versus objective (largest set possible).
+
+    Returns
+    -------
+    QUBO : dict
+       The QUBO with ground states corresponding to a maximum weighted independent set.
+    """
+
+    # empty QUBO for an empty graph
+    if not G:
+        return {}
+
+    N = len(G)
+
+    def index(a, b):
+        return (a)*N+(b)
+
+    ## Creating the QUBO
+    # Start with an empty QUBO
+    Q = {}
+    for i in range(N*N):
+        for j in range(N*N):
+            Q.update({(i,j): 0})
+
+    # Constraint that each row has exactly one 1
+    for v in range(N):
+        for j in range(N):
+            Q[(index(v,j), index(v,j))] += -1*lagrange
```

Comment:
because lagrange is defaulted to a float, float multiplication in python in is faster. So for instance
`Q[(index(v,j), index(v,j))] += -lagrange` or even `Q[(index(v,j), index(v,j))] -= lagrange`
and
`Q[(index(v,j), index(v,k))] += 2.0*lagrange`

the performance gain is tiny so it doesn't matter here really, but it's good practice :slightly_smiling_face: